### DISCUSSION TITLE
Next.js Auth Middleware + Server Actions

### DISCUSSION THREAD

**Discussion by acomanescu** [nextauthjs/next-auth#9783] [Help]
### Question 💬

I've been looking at the implementation of the auth middleware and I'm not sure how to properly handle the server actions.

Let's say what I have a single app, there are public pages (`/`, `/pricing`, etc.) and private pages (`/dashboard/*`).
I have an action that has to be protected (auth required), and it is shared between public pages and private pages.

When that action is called, it makes an http `POST` request with the same pathname as the page where it has been invoked. 

In this case, if I call it from a dashboard page (`/dashboard/profile`), the action makes a `POST` request with the pathname `/dashboard/profile`. Before reaching the action code, the middleware will intercept it and block it by redirecting if the request is not authenticated.

If I call the same action from `/profile` page, the action will make the request with pathname `/profile`. From my understanding, this will no longer be intercepted by the middleware and thus will not be blocked if not authenticated.

Server Actions might make specific middleware use cases invalid, such as this one due to how they are designed.

Thank you!

### How to reproduce ☕️

Create a shared Next.js Server Action between public and private pages that requires the request to be authenticated.
Add the auth middleware available.
Call the action from a page that's protected by the middleware then the same action from the public page.
The first call will be protected, but not the second.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by chungweileong94** [comment]
From what I know, server action share the same auth protection as the page you are calling the action from. 
So if you want to make a server action remain protected in any pages,  you can use `getServerSession` to check if the user is authenticate before you perform the action logic.

**Comment by sameedahri** [comment]
So we have to write getServerSession in every action, can I wrap my server action with some wrapper that can check for authentication? So I won't need to add it in all actions?

**Comment by chungweileong94** [comment]
> So we have to write getServerSession in every action, can I wrap my server action with some wrapper that can check for authentication? So I won't need to add it in all actions?
> 
> 

Yes, server action is more or less just a function, you can technically wrap it in anyway you want. I personally build my wrapper for this https://github.com/chungweileong94/server-act
